docs_src/design/ucr/Core-Data-Cache.md

```diff
@@ -0,0 +1,46 @@
+## Core Data Cache
+
+### Submitters
+Jim White (IOTech Systems)
+
+
+## Change Log
+
+### Market Segments
+Any/All
+
+### Motivation
+Reduction in the amount of data that is persisted at the edge.  Reduction in the amount of data sent to the north.  Reduction in the amount of data sent to edge analytics (rules engines, etc.).
+
+### Target Users
+- Device Manufacturer
+- Device Owner
+- Device User
+- Device Maintainer
+- Cloud Provider
+- Service Provider
+- Network Operator
+- Software Developer
+- Software Deployer
+- Software Integrator
+
+### Description
+In cases where there is a need to store data at the edge and that data is subsequently sent to the “north” (cloud or enterprise systems, rules engines, AI/ML, etc.), there may be a need to keep (persist) only the latest readings.  “Latest” should be configurable and defined by the user – allowing for a cache of so many readings for a particular device resource.  Queries of core data should also allow for requesting the “latest” N readings as well.
+
+For example, as a temperature sensor may report the current temperature (the device resource) very frequently (say once every 5 seconds), that data may only be sent to other services or systems every minute.  The user may wish to have only the last two readings persisted and subsequently sent north during the minute interval (batch and send).  Thus, core data becomes a kind of cache for a certain number of readings.
+
+### Existing solutions
+Today, core data will persist all data sent it.  The scheduler can be used to “clean” older data (data collected with a timestamp exceeding a specific timeframe).  But there is no way to only retain only X number or latest readings.  There is also no way to query for the “latest” readings.
+
+### Requirements
+
+- Cap the number of a device readings, per resource name, that are kept/persisted in core data.
```

Comment:
Is it a hard requirement to cap the number of readings to an exact number or is it sufficient to automatically purge entries beyond the limit? I'm asking because this will significantly affect the design and implementation. The former will require a FIFO persisted data storage, the latter can be achieved by just purging data at an optimal interval.

Reply:
Fair and good question.  I have been assuming that an exact number is required.  I will add this as a requirement but hope folks like @TomBrennan-Eaton can provide comment.
When you say computationally expensive - I am assuming because we need to do a delete or remove for each insert vs doing a delete once in a while which may not be for every insert.

Reply:
Interesting - I'd say that our interest is to ensure a minimum number of available entries; exceeding that would be fine (probably expected, for very recent readings). 
Is there an easy way to determine the "optimal interval" for purging? Could it be based on total count exceeding some high-water mark, as well as by time, in case we don't have a good means to determine an optimal time for every deployment?

Reply:
Yes, enforcing a hard cap is computationally more expensive especially if the data is persisted on disk. I don't know about Redis, but some time series databases accept delete requests and commit them in the background at a suitable interval. I guess it is best to leave that to the ADR and drop the hard requirement **if it there it is no use case for it.**

I think a configurable purging interval together with a high watermark as @TomBrennan-Eaton suggests could work.